post: nv12-vkms (#6)
Grillo-0/blog: _posts/2023-06-03-nv12-vkms.md

---
layout: post
title: "How it's Going: Adding NV12 support to VKMS"
date: 2023-06-03 12:30
category: kernel developing
tags: open-source graphics
---

> This work proposed by my mentor [Maira Canal][maira] is going more difficult than I

Reviewer: ```suggestion
This work proposed by my mentor [Maíra Canal][maira] is going more difficult than I
```

thought >W<.

# What is VKMS?

The [Virtual Kernel Mode Setting][vkms] is a virtual video driver. It is used for
testing the correct display of userspace programs without needing a physical

> monitor. This technique is useful for CI/CD for example.

Reviewer 2: ```suggestion
monitor. Virtual drivers are useful for automated testing.
```

The Color YCbCr has three color encoding standards, BT601, BT709, and BT2020,
besides that the YCbCr can occupy the full range of each byte it uses or a
limited range.

To tell what color encoding and range the driver support, we have to add the
[`drm_plane_create_color_properties()`][drm_plane_create_color_properties]

```diff
@@ -212,5 +212,14 @@ struct vkms_plane *vkms_plane_init(struct vkms_device *vkmsdev,
        drm_plane_create_rotation_property(&plane->base, DRM_MODE_ROTATE_0,
                                           DRM_MODE_ROTATE_MASK | DRM_MODE_REFLECT_MASK);

+       drm_plane_create_color_properties(&plane->base,
+                                         BIT(DRM_COLOR_YCBCR_BT601) |
+                                         BIT(DRM_COLOR_YCBCR_BT709) |
+                                         BIT(DRM_COLOR_YCBCR_BT2020),
+                                         BIT(DRM_COLOR_YCBCR_LIMITED_RANGE) |
+                                         BIT(DRM_COLOR_YCBCR_FULL_RANGE),
+                                         DRM_COLOR_YCBCR_BT601,
+                                         DRM_COLOR_YCBCR_FULL_RANGE);
+
        return plane;
 }
```

> The conversion code was taken from the [`tpg-core.c`][tpg-core.c], a virtual

Reviewer 2: this hyperlink seems broken

driver from the media subsystem that does those conversions in software as well.

As a side thought, maybe would be better to have those two subsystems use the
same code, maybe with a separate subsystem that handles color formats.

> The TPG code was changed to use the [`drm_fixed.h`][drm_fixed] operations, as to

Reviewer: ```suggestion
The TPG code was changed to use the [`drm_fixed.h`][drm_fixed] operations, to
```

intrinsically imperfect, or if the use of fixed-point operations is the culprit.
All I know is that certain colors are slightly off.

Luckily the IGT guys know about this issue, the way they overcome that is by
just checking the MSB bits of color values, basically rounding then. They do
that by passing a Gamma Look-Up Table (LUT) to the driver. But VKMS doesn't
support that.

# Before that, What is a Gamma LUT?

It is a look-up table in that the index' is the color and its value is the
result color.

This is the definition of a 1D table. You can have one for all the channels, so
the transformation is the same for all the color channels, or one for each
channel, so you can tweak each channel specifically.

![1D LUT representation](/assets/img/1d-gamma-lut.webp)

There is a more complex type of LUT, a 3D LUT. But this one I don't fully
understand and it's not needed. All I know is that you use the three color
channels at the same time for the index, one for each coordinate, and the value
that you get is a color. And besides that, you have to do interpolations.

![3D LUT representation](/assets/img/3d-gamma-lut.webp){: width="50%" }

> It image representation is a pretty cube :).

Reviewer: ```suggestion
Its visual representation is a pretty cube :).

```
